datafusion/physical-plan/src/joins/sort_merge_join.rs

```diff
@@ -914,23 +922,27 @@ impl SMJStream {
                     .disk_manager
                     .create_tmp_file("SortMergeJoinBuffered")?;
 
-                if let Some(batch) = buffered_batch.batch {
-                    spill_record_batches(
-                        vec![batch],
-                        spill_file.path().into(),
-                        Arc::clone(&self.buffered_schema),
-                    )?;
-                    buffered_batch.spill_file = Some(spill_file);
-                    buffered_batch.batch = None;
-
-                    // update metrics to register spill
-                    self.join_metrics.spill_count.add(1);
-                    self.join_metrics
-                        .spilled_bytes
-                        .add(buffered_batch.size_estimation);
-                    self.join_metrics.spilled_rows.add(buffered_batch.num_rows);
+                match buffered_batch.data {
+                    BufferedBatchData::InMemory(batch) => {
+                        spill_record_batches(
+                            vec![batch],
+                            spill_file.path().into(),
+                            Arc::clone(&self.buffered_schema),
+                        )?;
+                        buffered_batch.data = BufferedBatchData::Spilled(spill_file);
+
+                        // update metrics to register spill
+                        self.join_metrics.spill_count.add(1);
+                        self.join_metrics
+                            .spilled_bytes
+                            .add(buffered_batch.size_estimation);
+                        self.join_metrics.spilled_rows.add(buffered_batch.num_rows);
+                    }
+                    BufferedBatchData::Spilled(_) => {
```

Comment:
It isn't clear to me how this is prevented FWIW, but at least it now isn't silently ignored

Reply:
I think this is the error that @viirya is suggesting: https://github.com/apache/datafusion/pull/11218/files#r1682222150